### PR TITLE
Pass full path name to State.new for better errors

### PR DESCRIPTION
**Pull request by kbrock**
## Overview

`statelint` passes around the path to the states instead of just the name.

If we pass a full path to `State` and `ChoiceRule` we can improve validations

I want to go through each of the states to check all fields, but that just fixed up the existing errors and added a little bit of structure to help me fix ChoiceState

## Changes:

```diff
- State requires ErrorEquals
+ States.State.Retry.0 does not have required field "ErrorEquals"
- State requires ErrorEquals
+ States.State.Retry.0 does not have required field "ErrorEquals"
- Choice state must have "Choices"
+ States.Choice1 does not have required field "Choices"
- "Choices" must be a non-empty array
+ States.Choice1 field "Choices" must be a non-empty array
- "Choices" must be a non-empty array
+ States.Choice1 field "Choices" must be a non-empty array
- "Default" not in "States"
+ States.Choice1 field "Default" value "MissingState" is not found in "States"
- Missing field "States"
+ State Machine does not have required field "States"
- Invalid state type: [Invalid]
+ States.FirstState field "Type" value "Invalid" is not valid
- Missing field "StartAt"
+ State Machine does not have required field "StartAt"
- "StartAt" not in the "States" field
+ State Machine field "StartAt" value "Foo" is not found in "States"
- Missing "Type" field in state [FirstState]
+ States.FirstState does not have required field "Type"
- State name [#{truncated_state_name}] must be less than or equal to 80 characters
+ States field "Name" value "#{state_name}" must be less than or equal to 80 characters
- Invalid resource scheme [invalid]
+ States.FirstState field "Resource" value "invalid://foo" Invalid resource scheme [invalid]
```

### CODE REVIEW

**Comment by kbrock** [comment]
update:
- using `_full_name` for Retrier. Won't use variable until we start validating those fields

**Comment by kbrock** [comment]
update:
- rubocop

**Comment by kbrock** [comment]
@agrare This doesn't do much, it just passes around the name. Now that I know this is a viable direction, I have updated #209 to include the before and afters

**Comment by kbrock** [comment]
update:
- `s/Retrier/Catcher` -- a copy pasted error

File: lib/floe/workflow/states/task.rb [hunk 1]
```diff
@@ -20,8 +20,8 @@ def initialize(workflow, name, payload)
           @resource          = payload["Resource"]
           @runner            = Floe::Runner.for_resource(@resource)
           @timeout_seconds   = payload["TimeoutSeconds"]
-          @retry             = payload["Retry"].to_a.map { |retrier| Retrier.new(retrier) }
-          @catch             = payload["Catch"].to_a.map { |catcher| Catcher.new(catcher) }
+          @retry             = payload["Retry"].to_a.each_with_index.map { |retrier, i| Retrier.new(full_name + ["Retry", i.to_s], retrier) }
+          @catch             = payload["Catch"].to_a.each_with_index.map { |catcher, i| Catcher.new(full_name + ["Catch", i.to_s], catcher) }
```
Comment (Fryguy) [inline lib/floe/workflow/states/task.rb:24]:
```suggestion
          @retry             = payload["Retry"].to_a.each_with_index.map { |retrier, i| Retrier.new(full_name + ["Retry", i.to_s], retrier) }
          @catch             = payload["Catch"].to_a.each_with_index.map { |catcher, i| Catcher.new(full_name + ["Catch", i.to_s], catcher) }
```
```suggestion
          @retry             = payload["Retry"].to_a.map.with_index { |retrier, i| Retrier.new(full_name + ["Retry", i.to_s], retrier) }
          @catch             = payload["Catch"].to_a.map.with_index { |catcher, i| Catcher.new(full_name + ["Catch", i.to_s], catcher) }
```

**Comment by kbrock** [comment]
update:
- Added the validation messages changes from #209
- Cut down the common validation method to just a hand full of messages
- Did not add any new messages, but rather tried to minimize change

update:
- reduced flog score to make code climate happy

File: lib/floe/workflow/catcher.rb [hunk 2]
```diff
@@ -3,17 +3,19 @@
...
-        @payload = payload
+      def initialize(full_name, payload)
+        @full_name    = full_name
+        @payload      = payload
 
         @error_equals = payload["ErrorEquals"]
         @next         = payload["Next"]
         @result_path  = ReferencePath.new(payload.fetch("ResultPath", "$"))
-        raise Floe::InvalidWorkflowError, "State requires ErrorEquals" if !@error_equals.kind_of?(Array) || @error_equals.empty?
+        parse_error!("State requires ErrorEquals") if !@error_equals.kind_of?(Array) || @error_equals.empty?
```
Comment (kbrock) [inline lib/floe/workflow/catcher.rb:18]: ugh

```suggestion
        parser_missing_field!("ErrorEquals") if !@error_equals.kind_of?(Array) || @error_equals.empty?

        
```
```suggestion
        parse_error!("State requires ErrorEquals") if !@error_equals.kind_of?(Array) || @error_equals.empty?
```

**Comment by kbrock** [comment]
update:
- updated an `error!` to a `missing_field!` (same output, just using the correct method)

File: lib/floe/validation_mixin.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    private
+
+    def full_name_string
+      full_name.join(".")
+    end
+
+    module ClassMethods
+      def parser_error!(full_name, comment)
+        full_name = full_name.join(".") if full_name.kind_of?(Array)
```
Comment (agrare) [inline lib/floe/validation_mixin.rb:33]: Do we need this and also `#full_name_string`?  Could we enforce one or the other or are there some cases where we have to pass an array in?

File: lib/floe/workflow/state.rb [hunk 4]
```diff
@@ -4,32 +4,31 @@ module Floe
   class Workflow
     class State
       include Logging
+      include ValidationMixin
 
       class << self
-        def build!(workflow, name, payload)
+        def build!(workflow, full_name, payload)
```
Comment (agrare) [inline lib/floe/workflow/state.rb:10]: @kbrock do we care about name vs full_name?  Could we keep it as name but have it be the full path instead?

File: lib/floe/workflow/state.rb [hunk 5]
```diff
@@ -4,32 +4,31 @@ module Floe
...
           state_type = payload["Type"]
-          raise Floe::InvalidWorkflowError, "Missing \"Type\" field in state [#{name}]" if payload["Type"].nil?
+          parser_missing_field!(full_name, "Type") if payload["Type"].nil?
```
Comment (agrare) [inline lib/floe/workflow/state.rb:12]: Just naming stuff, but wonder if validate_missing_field! would be better?

Comment (kbrock) [reply lib/floe/workflow/state.rb:12]: This is not checking values but rather throwing an error.
It is essentially a wrapper that generates the proper text.

(I would prefer if all/most errors were of the form "$State field $Field $message", but had trouble keeping this language active vs passive, so I punted to `statelint` wording, and kept this as a place holder to fix.)

I will look for a better phrase

File: lib/floe/validation_mixin.rb (hunk 3 above)
Comment (kbrock) [reply lib/floe/validation_mixin.rb:33]: Good call. Workflow is a little funky with this, but the code seems fine dropping this extra concept

File: lib/floe/workflow/state.rb (hunk 4 above)
Comment (kbrock) [reply lib/floe/workflow/state.rb:10]: There are a few spots that reference the state's name and want the shorter name.
I've introduced short_name, so I guess we still needed this concept.

So we now have short_name, name (full name), and long_name (type: name)
Maybe we want to drop long_name at a later time

File: lib/floe/workflow/state.rb (hunk 5 above)
Comment (kbrock) [reply lib/floe/workflow/state.rb:12]: going to change to `missing_field_error!`

**Comment by kbrock** [comment]
update:
- renamed full_name to name (added short_name for those that required just the current name)
- updated error names to reflect they are throwing an error
- dropped full_name_str (it was handled by a single join)
- `wrap_parser_error` added localized rescues for parsing paths and the like - will circle back for this.
- default Workflow name to `"State Machine"` (so we could drop `full_name_str`)
- move some workflow specs to the state they are testing
- Add tests around Next, End, missing Task#Resource

update:
- cops

**Review by agrare — APPROVED** [review]

**Pull request merged**